lldb/source/Plugins/Process/Windows/Common/NativeRegisterContextWindows_arm64.cpp

```diff
@@ -143,8 +142,14 @@ NativeRegisterContextWindows::CreateHostNativeRegisterContextWindows(
 
 NativeRegisterContextWindows_arm64::NativeRegisterContextWindows_arm64(
     const ArchSpec &target_arch, NativeThreadProtocol &native_thread)
-    : NativeRegisterContextWindows(native_thread,
-                                   CreateRegisterInfoInterface(target_arch)) {}
+    : NativeRegisterContextRegisterInfo(
+          native_thread, CreateRegisterInfoInterface(target_arch)) {
+  // Currently, there is no API to query the maximum supported hardware
+  // breakpoints and watchpoints on Windows. The values set below are based
```

Comment:
On Windows for ARM64 or in general? I know this is a _arm64 class but just add "on Windows for arm64." in case someone like me is skimming through.

Reply:
>  // on tests conducted on Windows 11 with Snapdragon Elite X hardware.

As in this is what the Windows header had set those defines to, or this is what you were able to use?

I remember this style from my MIPS days but there you could write to the registers and read them back to tell which slots were active. I don't know if that's a viable option here.

Reply:
I dont think a API exists to query hardware breakpoint or watchpoint resources. On x86_64 as well winnt.h definitions of hardware breakpoint and watchpoints are used.

winnt.h defined the max hardware breakpoints to be 8 however in reality only 6 can be used successfully. Also for watchpoints it defines 2 arm64 hardware watchpoints but in reality only 1 can be used.

Reply:
Does Windows reject writes to the other breakpoints or just silently accepts them and only later you find out they don't work?

I'm wondering if we can read/modify/write the registers and see what sticks.

Of course the timing of that is interesting. We'd need a live process to do those writes in, and perhaps by that time we already needed to know the max breakpoint number for something else to work.

(last time I did this sort of thing it was a jtag debugger so we could inspect registers regardless of the debugee's state)

Reply:
In principle, there's nothing stopping you from launching a dummy process just for the purpose of testing the debug API. I noticed that this is what linux `strace` does when it tries to determine the host kernel capabilities. Windows makes this slightly trickier as you can't just call `fork` -- you need an actual executable -- but the same principle holds. (I'm not saying we need to do this here, I'm just throwing the idea out there.)

Reply:
In my testing  I was able to set and read the registers back successfully however the exception which should be generated as a result of setting those registers do not get triggered. So in theory you are able to set and read hardware breakpoint registers back with correct values. But there is some configuration internal to windows which probably does not happen until we land in user code for generating those exceptions.

Reply:
Also winnt.h defines wvr, wcr, bvr and bcr arrays which has size equal to ARM64_MAX_WATCHPOINTS or ARM64_MAX_BREAKPOINTS. So we should be able to read and write those arrays regardless of whether they eventually generate a watchpoint or breakpoint exception or not.

Reply:
Ok so the kernel structure is always there it just isn't going to write the values to hardware (or it does and the hardware ignores it, but same result either way).

It's a shame because this means an lldb-server built on WoA CpuX won't detect more or less points on CpuY (I think Jason already said this).

I also wouldn't say we need to use Pavel's idea right now. Another option would be to do it based on CPU ID if we have access to that in Windows.